### DISCUSSION TITLE
Consider Ruff ready for a 1.0.0 release?

### DISCUSSION THREAD

**Discussion by jonathanlermitage** [astral-sh/ruff#4505] [General]
I am wondering if it is time to publish the 1.0.0 release.  
From my point of view, Ruff is stable and mature. It works super well.  
But, I am facing this issue: some people consider 0.x or 0.0.x versions as non-stable. I faced this problem in my company: we cannot use Ruff in our projects because this is still a "beta" product. I strongly disagree, but hum, you know how it works in big companies 😅. The security guy observed that Ruff is still a beta product, so this is not stable, blablabla.  
So, this is why I suggest releasing the 1.0.0.  
Otherwise, do we have plans or milestones for a stable release?  
Thanks!

**Comment by tjkuson** [comment]
Looking at #1992, version one will only be released when the API has matured (for example, when it's more decoupled from flake8 and supports severity levels).

Regarding beta stigma, have you shown them how many large projects use ruff in production? It's got quite an impressive user base for a 0.0.x tool.

**Reply by jonathanlermitage** [reply]
Thanks for this explanation!  

> Regarding beta stigma, have you shown them how many large projects use ruff in production?

Yeah... I tried, without success. But it's okay, they will probably use Ruff _later_. The mentality in my company is very particular.

**Reply by simunic-cz** [reply]
@tjkuson
> ...many large projects use ruff in production...

That really means absolutely nothing - it's not a valid argument.

My company is operating in very tightly regulated field where security and compliance are strict. Even if the community is active and issues are dealt with quickly, you can't use an argument: „but they use it too“ during an audit. I mean, you can, but then you can't be surprised when, for instance, your cybersecurity insurance premiums go sky high - or when you eg. don't get/lose a contract with a client.

But ok, Ruff is an "internal tooling," not something deploy or "run in production." So, setting aside security and compliance, there is still the question of maintainability. I am a Platform engineer and one thing my team has to do every week is to update all dependencies. We have some semi-automated pipelines for this and they prepare a PR every week with all minor versions updated. But when a breaking change occurs, we still have to deal with it.

Ruff does NOT have a stable API, that's written literary on official Ruff website: https://docs.astral.sh/ruff/versioning
Ruff is NOT using semantic versioning. Any minor version increase can contain a breaking change. Now, if I have to spend time dealing with new breaking changes on effectively weekly basis, then I absolutely do not care about how fast Ruff is compared to an alternative.

**Unstable API = high maintenance. I have many much more pressing issues to deal with then resolving breaking changes every week just because Ruff is using, and I'm quoting: „custom versioning scheme.“**

PS: In my experience, large quantity of commonly used Python libraries are a real pain in the butt to work with simply because their maintainers are either unwilling or don't know how to use semantic versioning.

**PPS: I'm really looking forward for Ruff to become stable. It looks awesome and I'm sure it will make our lives easier and development much more pleasant. But I'm afraid that can't happen on a wider scope until Ruff actually becomes stable.**

**Reply by tjkuson** [reply]
I'm not a part of the Astral team, just speaking as a user and contributor (though I haven't contributed much of late).

> That really means absolutely nothing - it's not a valid argument.

Well, it means _something_. For example, when lots of people use a tool, issues tend to be discovered quickly and solutions in turn. This might not be convincing in your situation, but I think the incredible uptake of Ruff is a compelling argument for many, so it seems unfair to suggest it 'means nothing' or is invalid.

Additionally, I think this argument has strengthened in the nearly three years after I posted that comment, with more and more people adopting the tool.

> My company is operating in very tightly regulated field where security and compliance are strict. Even if the community is active and issues are dealt with quickly, you can't use an argument: „but they use it too“ during an audit. I mean, you can, but then you can't be surprised when, for instance, your cybersecurity insurance premiums go sky high - or when you eg. don't get/lose a contract with a client.

Candidly, I also work is a field with fairly strict requirements and have not encountered this problem. Libraries like httpx and Fast API are production-ready and ZeroVer, and I don't recall issues with those either. However, this does sound quite frustrating for you, and hope you can find a solution soon!

> We have some semi-automated pipelines for this and they prepare a PR every week with all minor versions updated.

Have you considered only automatically updating Ruff in patch updates only? There shouldn't be disruptive changes there.

> Now, if I have to spend time dealing with new breaking changes on effectively weekly basis, then I absolutely do not care about how fast Ruff is compared to an alternative.

To be clear, Ruff does not have weekly breaking changes. I know of very few breaking changes in Ruff in the past couple of years, and all I can recall have been flagged quite explicitly with migration guides. I think the Astral team does a pretty great job at making breaking changes infrequent and easy to deal with, personally.

A couple of thoughts on SemVer in general:

- For me, the purpose of a versioning system is communication. The system Astral chose (quite deliberately, I hazard, and not due to an ignorance of SemVer) has been very effective in communicating changes. So long as the versioning system is clear and explicit, I don't really mind what Ruff goes with (especially when we have other systems like CalVar, and the custom versioning system that the Python interpreter itself uses).
- Due to the nature of linters, even bug patch fixes can suddenly block CI. By that metric, any change to a linter can be considered, in practice, 'breaking'. It's unclear to me that even if Ruff were following SemVar and were v1 that it would solve your problems. I find the argument in the Biome docs compelling ([ref](https://biomejs.dev/internals/versioning/)): they recommend pinning linters to an exact version.

Regardless, I hope you and your team will be able to adopt Ruff soon. It's a pretty neat tool

**Reply by jonathanlermitage** [reply]
@patrik-simunic-cz 
Different companies, different people, different experiences.. you know what it means ^_^.

I don't think the problem is on the Ruff side. I mean, if the security guy of Big-Company-123 can't understand that Ruff is ready for production just because it's not labelled 1.0.0+, we can do nothing for him. I know, that sucks. Especially when we cant discuss with the security guy (this is what happened to me).

For the breaking changes, did you face so many compatibility issues? Seriously ? Also, could you simply stick to a particular version (at least for some time), then focus on more important things? I have a completely different experience, but I don't want to generalize. You may have different/exotic configuration files, which could explain your issues, I don't know.

Ruff *could* be labelled as 1.0.0, especially if it's using a custom version scheme. This would break nothing. This way, the security guy would say "yeah, it's stable, I approve!". Meanwhile, I think it's very important to respect the decision of Ruff's developers. The problem is not on the Ruff side, it's their (the security guys) fault if they're too stupid.  
I think the main question is: *should Ruff adapt to stupid people?* (actually, this is why I opened this discussion) I can't say, and this is not my role. Some times, I read "people who contribute nothing don't get a seat at the table", and I agree. I contribute to many open and closed source projects, but not to Ruff, so I can only suggest things, nothing more, and here, they explained why Ruff is still not labelled 1.0.0.  
If I really need a 1.0.0, I can still fork Ruff, and create a 1.0.0 release just for that. Yeah, I can be an idiot too 😁 (I left my job before doing that, but I was very close to doing so, just to see).
